src/setupCleanup.ts

```diff
@@ -0,0 +1,12 @@
+import { History } from "history";
+import { Store } from 'redux'
+import { TempDataRootState } from "./types";
+import { cleanupTempData } from "./actions";
+
+const setupCleanup = (history: History, store: Store<TempDataRootState>) => {
+  return history.listen((location) => {
```

Comment:
We are ignoring `unregister` history. I'm not sure about it.
we can just rely on the `React Router`🤔

Reply:
First, I tried to implement this logic as a middleware but find out that the middleware logic runs with every dispatch action and it's not a good option because I have to register it once.
In addition, the middleware does not run at the program init phase and waiting for the first action. So I think it is not suitable for subscribing to events.
Also, we need the store variable in the middleware body to be able to dispatch our onChange events.

Reply:
The idea of setting or running something after store creation comes from sagaMiddleware.run() method.
The connected-react-router has a similar issue and implements ConnectedRouter component that wraps the App. They handle subscribe and unsubscribe to history.listen in the lifecycles of this component.

Reply:
About the unregister, I think it is needed in a rare case and in most scenario, it just needs to register once. So the unregister is returned by the setupCleanup function and the user can unregister it by its own logic.
I think this method is like the sagaMiddleware.run(). Have you ever stop the saga after running it?

Reply:
You are right. we can continue for now. I just worried about hard steps to setup but it is ok for now.